### DISCUSSION TITLE
Please add support for GC9D01N_40*160 LCD

### DISCUSSION THREAD

**Discussion by iggymayer** [Bodmer/TFT_eSPI#2638] [New processor/display support or capability request]
Datasheet:
https://mega.nz/folder/VNMkzKiR#A7V1B4k2vza7yTO_WvD5-Q

Thank you!

**Comment by iggymayer** [comment]
New files has been uploaded:   TTH089HVG-01.pdf & GC9D01N_init.txt

**Comment by Bodmer** [comment]
Post web links to vendor pages for the screens.

**Comment by iggymayer** [comment]
https://www.panelook.cn/QT099-0416THBIG01-detail_cn_158790.html
https://www.panelook.cn/TTH089HVG-01-detail_cn_159005.html
https://item.taobao.com/item.htm?spm=a230r.1.14.1.79b84a04eGyh5m&id=677743417274&ns=1&abbucket=4#detail
https://item.taobao.com/item.htm?spm=a230r.1.14.33.79b84a04eGyh5m&id=686026085866&ns=1&abbucket=4#detail

**Comment by Bodmer** [comment]
Since I do not have one of those displays it would not be possible to test any updates. Do you have a display connected up and running with code? If so which processor are you using?

**Reply by iggymayer** [reply]
ESP32-C3
Replaced the GC9A01 initialization code with GC9D01N initialization code, it can be displayed but the display color is incorrect, and the picture flashes quickly (or jitter?). GC9A01_Defines.h & GC9A01_Rotation.h is not modified,It should be that the relevant register is set incorrectly.

**Reply by Bodmer** [reply]
Post your modified init file so I can check it.

**Reply by iggymayer** [reply]
{
  writecommand(0xFE);
  writecommand(0xEF);
  // 内部寄存器使能80~8Fh
  writecommand(0x80);
  writedata(0xFF);

  writecommand(0x81);
  writedata(0xFF);

  writecommand(0x82);
  writedata(0xFF);

  writecommand(0x83);
  writedata(0xFF);

  writecommand(0x84);
  writedata(0xFF);

  writecommand(0x85);
  writedata(0xFF);

  writecommand(0x86);
  writedata(0xFF);

  writecommand(0x87);
  writedata(0xFF);

  writecommand(0x88);
  writedata(0xFF);

  writecommand(0x89);
  writedata(0xFF);

  writecommand(0x8A);
  writedata(0xFF);

  writecommand(0x8B);
  writedata(0xFF);

  writecommand(0x8C);
  writedata(0xFF);

  writecommand(0x8D);
  writedata(0xFF);

  writecommand(0x8E);
  writedata(0xFF);

  writecommand(0x8F);
  writedata(0xFF);

  writecommand(0x3A);
  writedata(0x05);
  // 设置翻转方式
  writecommand(0xEC);
  writedata(0x11);

  // VGL大小
  writecommand(0x7E);
  writedata(0x7a);

  // 修改帧频
  writecommand(0x74);
  writedata(0x02);
  writedata(0x0E);
  writedata(0x00);
  writedata(0x00);
  writedata(0x28);
  writedata(0x00);
  writedata(0x00);

  // 内部电压调整
  writecommand(0x98);
  writedata(0x3E);
  writecommand(0x99);
  writedata(0x3E);

  // 内部porch设置
  writecommand(0xB5);
  writedata(0x0E);
  writedata(0x0E);

  // gip timing start
  writecommand(0x60);
  writedata(0x38);
  writedata(0x09);
  writedata(0x6D);
  writedata(0x67);

  writecommand(0x63);
  writedata(0x38);
  writedata(0xAD);
  writedata(0x6D);
  writedata(0x67);
  writedata(0x05);

  writecommand(0x64);
  writedata(0x38);
  writedata(0x0B);
  writedata(0x70);
  writedata(0xAB);
  writedata(0x6D);
  writedata(0x67);

  writecommand(0x66);
  writedata(0x38);
  writedata(0x0F);
  writedata(0x70);
  writedata(0xAF);
  writedata(0x6d);
  writedata(0x67);

  writecommand(0x6A);
  writedata(0x00);
  writedata(0x00);

  writecommand(0x68);
  writedata(0x3B);
  writedata(0x08);
  writedata(0x04);
  writedata(0x00);
  writedata(0x04);
  writedata(0x64);
  writedata(0x67);

  writecommand(0x6C);
  writedata(0x22);
  writedata(0x02);
  writedata(0x22);
  writedata(0x02);
  writedata(0x22);
  writedata(0x22);
  writedata(0x50);

  writecommand(0x6E);
  writedata(0x00);
  writedata(0x00);
  writedata(0x00);
  writedata(0x00);
  writedata(0x07);
  writedata(0x01);
  writedata(0x13);
  writedata(0x11);

  writedata(0x0B);
  writedata(0x09);
  writedata(0x16);
  writedata(0x15);
  writedata(0x1D);
  writedata(0x1E);
  writedata(0x00);
  writedata(0x00);

  writedata(0x00);
  writedata(0x00);
  writedata(0x1E);

  writedata(0x1D);
  writedata(0x15);
  writedata(0x16);
  writedata(0x0A);
  writedata(0x0C);

  writedata(0x12);
  writedata(0x14);
  writedata(0x02);
  writedata(0x08);
  writedata(0x00);
  writedata(0x00);
  writedata(0x00);
  writedata(0x00);

  // gip timing end

  // 内部电压设定开始
  writecommand(0xA9);
  writedata(0x1B);

  writecommand(0xA8);
  writedata(0x6B);

  writecommand(0xA8);
  writedata(0x6D);

  writecommand(0xA7);
  writedata(0x40);

  writecommand(0xAD);
  writedata(0x47);

  writecommand(0xAF);
  writedata(0x73);

  writecommand(0xAF);
  writedata(0x73);

  writecommand(0xAC);
  writedata(0x44);

  writecommand(0xA3);
  writedata(0x6C);

  writecommand(0xCB);
  writedata(0x00);

  writecommand(0xCD);
  writedata(0x22);

  writecommand(0xC2);
  writedata(0x10);

  writecommand(0xC5);
  writedata(0x00);

  writecommand(0xC6);
  writedata(0x0E);

  writecommand(0xC7);
  writedata(0x1f);

  writecommand(0xC8);
  writedata(0x0E);
  // 内部电压设定结束

  // 选择single gate mode
  writecommand(0xbf);
  writedata(0x00);
  // SOU相关调整
  writecommand(0xF9);
  writedata(0x20);

  // vreg电压调整
  writecommand(0x9b);
  writedata(0x3b);
  writecommand(0x93);
  writedata(0x33);
  writedata(0x7f);
  writedata(0x00);

  // VGH/VGL CLK调整70，71h
  writecommand(0x70);
  writedata(0x0E);
  writedata(0x0f);
  writedata(0x03);
  writedata(0x0e);
  writedata(0x0f);
  writedata(0x03);

  writecommand(0x71);
  writedata(0x0e);
  writedata(0x16);
  writedata(0x03);

  // 内部电压调整
  writecommand(0x91);
  writedata(0x0e);
  writedata(0x09);

  // vreg电压调整
  writecommand(0xc3);
  writedata(0x2c);
  writecommand(0xc4);
  writedata(0x1a);

  // gamma F1~F3h
  writecommand(0xf0);
  writedata(0x51);
  writedata(0x13);
  writedata(0x0c);
  writedata(0x06);
  writedata(0x00);
  writedata(0x2f);

  writecommand(0xf2);
  writedata(0x51);
  writedata(0x13);
  writedata(0x0c);
  writedata(0x06);
  writedata(0x00);
  writedata(0x33);

  writecommand(0xf1);
  writedata(0x3c);
  writedata(0x94);
  writedata(0x4f);
  writedata(0x33);
  writedata(0x34);
  writedata(0xCf);

  writecommand(0xf3);
  writedata(0x4d);
  writedata(0x94);
  writedata(0x4f);
  writedata(0x33);
  writedata(0x34);
  writedata(0xCf);

  writecommand(0x36);
  writedata(0x40);
  writecommand(0x11);
  delay(200);
  writecommand(0x29);
  writecommand(0x2C);
}

**Reply by Bodmer** [reply]
Run Read_User_Setup and post the serial output here so I can see your hardware configuration.

It appears the display is not being configured properly. Perhaps it needs a lower SPI clock speed?

**Reply by iggymayer** [reply]
Set the SPI frequency to 40 or 27, and the display effect is still the same.

[code]
TFT_eSPI ver = 2.5.23
Processor    = ESP32
Frequency    = 160MHz
Transactions = Yes
Interface    = SPI
Display driver = 9A01
Display width  = 40
Display height = 160

MOSI    = GPIO 6
MISO    = GPIO 6
SCK     = GPIO 4
TFT_DC   = GPIO 3
TFT_RST  = GPIO 1
TOUCH_CS = GPIO -21

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
Touch SPI frequency   = 2.50
[/code]

**Reply by Bodmer** [reply]
Do you have the display working with other software?

**Reply by iggymayer** [reply]
vscode&platformio

**Reply by iggymayer** [reply]
Rotation.h file TFT_MAD_BGR modified to TFT_MAD_RGB, the color anomaly problem has been resolved.

  rotation = m % 4;

  writecommand(TFT_MADCTL);
  switch (rotation) {
    case 0: // Portrait
      writedata(TFT_MAD_RGB);
      _width  = _init_width;
      _height = _init_height;
      break;
    case 1: // Landscape (Portrait + 90)
      writedata(TFT_MAD_MX | TFT_MAD_MV | TFT_MAD_RGB);
      _width  = _init_height;
      _height = _init_width;
      break;
    case 2: // Inverter portrait
      writedata(TFT_MAD_MX | TFT_MAD_MY | TFT_MAD_RGB);
      _width  = _init_width;
      _height = _init_height;
      break;
    case 3: // Inverted landscape
      writedata(TFT_MAD_MV | TFT_MAD_MY | TFT_MAD_RGB);
      _width  = _init_height;
      _height = _init_width;
      break;
  }

**Reply by Bodmer** [reply]
Great, so the flickering is gone too, and it is all working OK?

**Reply by iggymayer** [reply]
There is also a problem that there are several pixels in the corners of the screen that show other colors all the time, but the screen is normal (all pixels have no defective pixel).

**Reply by iggymayer** [reply]
![image](https://github.com/Bodmer/TFT_eSPI/assets/29495646/7f3fbee5-96d5-4ead-9223-93477191808c)

fillRect(1, 1, 38, 158, TFT_RED)
https://github.com/Xinyuan-LilyGO/T-0.99TFT
Use this library to display normally.

**Comment by Shamsyrup** [comment]
> ![image](https://user-images.githubusercontent.com/29495646/239904856-7f3fbee5-96d5-4ead-9223-93477191808c.jpeg)
> 
> fillRect(1, 1, 38, 158, TFT_RED) https://github.com/Xinyuan-LilyGO/T-0.99TFT Use this library to display normally.

Hi could you share your code and modified files ? I Got the same display and unable to display anything.

**Reply by iggymayer** [reply]
init code:
https://mega.nz/folder/VNMkzKiR#A7V1B4k2vza7yTO_WvD5-Q

rotation code:
https://github.com/Bodmer/TFT_eSPI/discussions/2638#discussioncomment-5966688

There is currently a defect that cannot be solved and does not display perfectly.

⬇Use this library to display normally.
https://github.com/Xinyuan-LilyGO/T-0.99TFT

If it still doesn't display, replace the screen or check the hardware design.
